genkit-tools/telemetry-server/src/localFileTraceStore.ts

```diff
@@ -47,32 +48,35 @@ export class LocalFileTraceStore implements TraceStore {
     filters?: Record<string, string>;
     storeRoot?: string;
     indexRoot?: string;
-    useIndex?: boolean;
   }) {
     this.storeRoot =
       options?.storeRoot ?? path.resolve(process.cwd(), `.genkit/traces`);
-    // Index is ephemeral, gets rebuilt on every restart. It's done for simplicity
-    // for now, so that we can tweak index format without having to worry about
-    // backwards compatibility. In the future we can keep indexes around between
-    // restarts and only rebuild if we detect corruption.
-    this.indexRoot = path.resolve(
-      os.tmpdir(),
-      `./genkit-trace-index-${Date.now()}-${Math.random()}`
-    );
     fs.mkdirSync(this.storeRoot, { recursive: true });
+    this.indexRoot =
+      options?.indexRoot ?? path.resolve(process.cwd(), `.genkit/traces_idx`);
+    fs.mkdirSync(this.indexRoot, { recursive: true });
     console.info(
       `[Telemetry Server] initialized local file trace store at root: ${this.storeRoot}`
     );
     this.filters = options?.filters ?? LocalFileTraceStore.defaultFilters;
-    if (options?.useIndex) {
-      this.index = new Index(this.indexRoot);
-    }
+    this.index = new Index(this.indexRoot);
   }
 
   async init() {
-    if (!this.index) {
-      return;
+    const metadata = this.index.getMetadata();
+    // if the metadata file doesn't exist or it was for the older version or if
+    // there are too many index files we recreate the index.
+    if (
+      !metadata ||
+      metadata.version !== currentVersion ||
+      this.index.listIndexFiles().length > MAX_INDEX_FILES
```

Comment:
Why not delete the oldest files in this case?

Reply:
Because an index file may contain 1 trace or 1000+ traces. For example what might happen is that the latest 9 files have 1 trace each, but the 10th has 100. If we just delete the oldest one, the index will only have 9 traces, but missing the rest. There are weirder corner cases when there are concurrent telemetry server instances running at the same time...